jerry-core/api/jerry.c

```diff
@@ -207,22 +197,33 @@ jerry_cleanup (void)
 
   jmem_finalize ();
   jerry_make_api_unavailable ();
-
-  if (JERRY_CONTEXT (user_context_deinit_cb))
-  {
-    JERRY_CONTEXT (user_context_deinit_cb) (JERRY_CONTEXT (user_context_p));
-  }
 } /* jerry_cleanup */
 
 /**
  * Retrieve user context.
  *
- * @return the user-provided context-specific pointer
+ * @param manager_p pointer to the manager whose user context item should be returned.
+ *
+ * @return the user-provided context-specific pointer for the given manager, creating such a pointer if none was found.
  */
-void *
-jerry_get_user_context (void)
+jerry_user_context_item_t *
+jerry_get_user_context (const jerry_user_context_manager_t *manager_p)
 {
-  return JERRY_CONTEXT (user_context_p);
+  jerry_user_context_item_t *item_p;
+
+  for (item_p = JERRY_CONTEXT (user_context_p); item_p; item_p = item_p->next_p)
+  {
+    if (item_p->manager_p == manager_p)
+    {
+      return item_p;
+    }
+  }
+
```

Comment:
I'm mostly concerned about type safety here. When we first designed the user context interface, we made init/deinit safe by requiring a pair of functions. That way, we could ensure that the right deinit was called every time. However, we left the get () completely separate. Thus, I realized that the following problem might happen:
```c
jerry_init_with_user_context(JERRY_INIT_EMPTY, jerry_foo_new, jerry_foo_free);
...
/* A mistake in the code - calls jerryx_bar_do_something_useful()
 * which calls jerry_get_user_context () assuming the context was
 * initialized with jerry_bar_new()/jerry_bar_free() */
jerryx_bar_do_something_useful();
```

Additionally, I also realized that jerryx_context has another shortcoming. Users of jerryx_context must call `jerryx_context_get_slot ()` instead of `jerry_get_user_context ()` to retrieve the context pointer. But how does context user `foo` know that the context was initialized with
```c
jerry_init_with_user_context (JERRY_INIT_EMPTY, jerryx_user_context_init, jerryx_user_context_deinit);
```
and not with
```c
jerry_init_with_user_context (JERRY_INIT_EMPTY, jerryx_foo_init, jerryx_foo_deinit);
```
I'd solved this latter problem with  -DJERRYX_HAVE_CONTEXT at build time and an ifdef that calls `jerry_get_user_context ()` if the flag is absent, and `jerryx_context_get_slot ()` if the flag is absent.

But this means that it does the same thing for every context.

And besides, it is only when an extension calls `jerry_get_user_context ()` that you can be certain that the extension needs the context.